File: android/app/src/main/assets/js/ting.js
/* 安康159 - 听牌/进张分析(JS版)
 * 翻译自 backend/rules/ting.py
 */

function waitingTiles(counts13) {
  const waits = [];
  for (let t = 0; t < TILE_COUNT; t++) {
    if (counts13[t] >= 4) continue;
    counts13[t]++;
    if (isWin(counts13)) waits.push(t);
    counts13[t]--;
  }
  return waits;
}

function isTing(counts13) {
  return waitingTiles(counts13).length > 0;
}

/**
 * 返回每张进张能将向听数降到多少(只列能降低向听数的进张)
 * 对齐 backend/rules/ting.py 的 useful_draws
 * @returns {Map<number, number>} tile -> 降后的向听数
 */
function usefulDraws(counts13) {
  const base = shanten(counts13);
  const result = new Map();
  for (let t = 0; t < TILE_COUNT; t++) {
    if (counts13[t] >= 4) continue;
    counts13[t]++;
    const s = shanten(counts13);
    counts13[t]--;
    if (s < base) result.set(t, s);
  }
  return result;
}

function discardOptions(counts14) {
  const options = [];
  for (let t = 0; t < TILE_COUNT; t++) {
    if (counts14[t] <= 0) continue;
    counts14[t]--;
    const s = shanten(counts14);
    const waits = s === 0 ? waitingTiles(counts14) : [];
    const waitCount = waits.reduce((sum, w) => sum + (4 - counts14[w]), 0);
    options.push({ tile: t, shanten: s, waits, wait_count: waitCount });
    counts14[t]++;
  }
  return options;
}

File: backend/rules/ting.py
"""安康159 - 听牌/进张分析

给定手牌计数(13张,未摸牌状态),枚举所有可能的进张,
判断摸到每张牌后是否能胡(听牌判断),以及打出每张牌后的听口。
"""

from .tiles import TILE_COUNT, tile_rank, is_159
from .win import is_win, shanten

RED = 27


def waiting_tiles(counts13: list[int]) -> list[int]:
    """13张手牌,返回所有能胡的进张列表(听口)"""
    waits = []
    for t in range(TILE_COUNT):
        if counts13[t] >= 4:
            continue
        counts13[t] += 1
        if is_win(counts13):
            waits.append(t)
        counts13[t] -= 1
    return waits


def is_ting(counts13: list[int]) -> bool:
    return len(waiting_tiles(counts13)) > 0


def useful_draws(counts13: list[int]) -> dict[int, int]:
    """返回每张进张能将向听数降到多少(只列能降低向听数的进张)"""
    base = shanten(counts13)
    result = {}
    for t in range(TILE_COUNT):
        if counts13[t] >= 4:
            continue
        counts13[t] += 1
        s = shanten(counts13)
        counts13[t] -= 1
        if s < base:
            result[t] = s
    return result


def discard_options(counts14: list[int]) -> list[dict]:
    """14张(摸牌后),枚举打出每张牌后的听牌情况。
    返回 [{'tile': t, 'shanten': s, 'waits': [...], 'wait_count': n}]"""
    options = []
    for t in range(TILE_COUNT):
        if counts14[t] <= 0:
            continue
        counts14[t] -= 1
        s = shanten(counts14)
        waits = waiting_tiles(counts14) if s == 0 else []
        # 剩余可摸的进张数(粗略: 4 - 手里已有)
        wait_count = int(sum(4 - counts14[w] for w in waits))
        options.append({
            "tile": t,
            "shanten": s,
            "waits": waits,
            "wait_count": wait_count,
        })
        counts14[t] += 1
    return options
